lib/PrintAsClang/PrimitiveTypeMapping.cpp

```diff
@@ -19,29 +19,78 @@
 
 using namespace swift;
 
+/// Find the implementation of the named type in the named module if loaded.
+static TypeDecl *findTypeInModuleByName(ASTContext &ctx,
+                                        Identifier moduleName,
+                                        Identifier typeName) {
+  auto module = ctx.getLoadedModule(moduleName);
+  if (!module)
+    return nullptr;
+
+  // Find all of the declarations with this name in the Swift module.
+  SmallVector<ValueDecl *, 1> results;
+  module->lookupValue(typeName, NLKind::UnqualifiedLookup, results);
+  for (auto result : results) {
+    if (auto nominal = dyn_cast<NominalTypeDecl>(result))
+      return nominal;
+
+    if (auto typealias = dyn_cast<TypeAliasDecl>(result))
+      return typealias;
+  }
+
+  return nullptr;
+}
+
 void PrimitiveTypeMapping::initialize(ASTContext &ctx) {
   assert(mappedTypeNames.empty() && "expected empty type map");
+
+  auto addMappedType = [&](Identifier moduleName, Identifier typeName,
+                           FullClangTypeInfo info) {
+    auto decl = findTypeInModuleByName(ctx, moduleName, typeName);
+    if (!decl)
+      return;
+
+    // Always map a direct definition match. Either the nominal decl or the
+    // typealias itself.
+    mappedTypeNames[decl] = info;
+
+    // If the underlying type of a typealias doesn't have a type, set it here.
+    // This aims to reproduce the typealias behavior from BuiltinMappedTypes.
+    if (auto typealias = dyn_cast<TypeAliasDecl>(decl)) {
+      auto underlying = typealias->getDeclaredInterfaceType()->getAnyNominal();
+      if (underlying && !mappedTypeNames.contains(underlying))
+        mappedTypeNames[underlying] = info;
```

Comment:
Should we have some logic here to double check we would add the same ClangTypeInfo for the same type?

Reply:
We may want to check that the size is the same, but we currently allow the printed type to be different. The test has the example of `CChar32` and `CWideChar`, both are aliases to `Unicode.Scalar` on macOS and have the same size but different print types: `char32_t` and `wchar_t`. We use the first one to be defined for the underlying type.

This logic is a somewhat backwards as it aligns with the one on the side of BuiltinMappedTypes. Ideally I would like to set the ClangTypeInfo of the underlying type and have aliases define a more precise version. But as long as BuiltinMappedTypes looks through typealiases we may yet again define an underlying type unwittingly and not have a matching definition here in PrimitiveTypeMapping.

Reply:
I see. I am also a bit unsure if this logic will work well with SIMD types. E.g., `simd.float4` might be a typealias for `SIMD4<Float>`.  If we get the underlying nominal declaration we might end up with the generic `SIMD4<T>`. We definitely don't want that to be part of the mapped types.

Reply:
It does look like `float4` is a typealias of `SIMD4<Float>`. Neither of the `float4` or the `SIMD<>` are listed on the side of BuiltinMappedTypes so they don't rely on the typealias logic. I'm adding an opt-out for these types to avoid associating the information down to the `SIMD<>` types.

Note that the `SIMD<>` style is not currently considered as representable in C/Objective-C so they should already be prevented from being printed in the compatibility header.

Did you have concerns beyond the SIMD case?

Reply:
About checking the size, we don’t have access to that information readily here since we work with C type names as strings and their size is platform-specific. We may want to implement this check on the side of the builtin types definition instead. I think we have the size information there and if there’s a mismatch we’d have bigger problems at that level.